### ISSUE TITLE
Sectioning algorithm doesn’t take into account neighbouring headings

### ISSUE THREAD

**Issue by pepelsbey** [mdn/fred#1113]
## Description

While working on a fix for https://github.com/mdn/fred/issues/690 and testing how adjacent headings are rendered, I discovered an issue when `h2` is **always** split into its own section.

## Example

[GLSL shaders](https://developer.mozilla.org/en-US/docs/Games/Techniques/3D_on_the_web/GLSL_Shaders#final_code)

```md
## Final code

### HTML

<!-- Content -->
```

Becomes two separate sections:

```html
<section class="content-section" aria-labelledby="final_code">
  <h2 id="final_code" class="heading">
    <a class="heading-anchor" href="#final_code">Final code</a>
  </h2>
</section>
<section class="content-section" aria-labelledby="html">
  <h3 id="html" class="heading">
    <a class="heading-anchor" href="#html">HTML</a>
  </h3>
  <!-- Content -->
</section>
```

## More details

It’s all fine when there’s content between headings:

```md
## Heading level two

I'm baby kale chips slow-carb cardigan gochujang.

### Heading level three

Subway tile DSA kinfolk cray meggings. 
```

But it starts failing when headings are adjacent:

```md
## Heading level two

### Heading level three

Subway tile DSA kinfolk cray meggings. 
```

…which creates two separate sections: one with just h2. Which doesn’t make sense.

```html
<section class="content-section" aria-labelledby="heading_level_two_2">
  <h2 id="heading_level_two_2" class="heading">
    <a class="heading-anchor" href="#heading_level_two_2">
      Heading level two
    </a>
  </h2>
</section>

<section class="content-section" aria-labelledby="heading_level_three_2">
  <h3 id="heading_level_three_2" class="heading">
    <a class="heading-anchor" href="#heading_level_three_2">
      Heading level three
    </a>
  </h3>
  <p>Subway tile DSA kinfolk cray meggings.</p>
</section>
```

Unlike when `h3` is followed by `h4`.

```md
### Heading level three

#### Heading level four
```

That gives proper section with adjacent headings:

```html
<section class="content-section" aria-labelledby="heading_level_three_3">
  <h3 id="heading_level_three_3" class="heading">
    <a class="heading-anchor" href="#heading_level_three_3">
      Heading level three
    </a>
  </h3>
  <h4 id="heading_level_four_2">
    Heading level four
  </h4>
</section>
```

<details>
<summary>Full Markdown content structure for testing</summary>

```md
## Heading level two

I'm baby kale chips slow-carb cardigan gochujang, godard austin selvage master cleanse. Same four dollar toast af kinfolk, neutral milk hotel JOMO shabby chic sustainable hashtag ethical DSA raclette farm-to-table. Salvia chillwave vaporware Brooklyn. Gastropub DIY vinyl ascot stumptown williamsburg helvetica polaroid pitchfork tilde jawn kogi schlitz. Fam fanny pack glossier shoreditch occupy direct trade skateboard williamsburg godard scenester edison bulb salvia. Food truck viral meggings aesthetic, vaporware artisan organic twee before they sold out fingerstache mumblecore.

### Heading level three

Subway tile DSA kinfolk cray meggings. Cupping beard skateboard, iPhone praxis grailed lyft. Fit butcher narwhal, brunch farm-to-table intelligentsia hell of grailed four dollar toast tofu. Biodiesel iceland stumptown lyft. La croix 8-bit flexitarian offal, before they sold out master cleanse mumblecore bicycle rights artisan ascot franzen. Tonx health goth adaptogen enamel pin vinyl farm-to-table normcore flexitarian put a bird on it swag.

#### Heading level four

Gatekeep cardigan marfa sus cred, glossier aesthetic chia palo santo four loko poutine. Flannel venmo narwhal, chicharrones wolf distillery kale chips man braid post-ironic bespoke tilde gorpcore church-key. Fingerstache gastropub bruh, food truck next level activated charcoal gorpcore viral. Hoodie intelligentsia tonx skateboard green juice, mumblecore affogato pok pok. Keffiyeh jianbing tofu, williamsburg vaporware distillery neutra cred seitan.

##### Heading level five

VHS schlitz art party neutra letterpress. Small batch franzen typewriter big mood. Meditation stumptown disrupt, iceland kale chips paleo fit trust fund prism banh mi vice cronut vaporware schlitz raw denim. Cronut sartorial PBR&B listicle celiac farm-to-table kale chips pour-over vice pop-up leggings ascot tumblr. 8-bit man braid slow-carb banh mi organic, jawn helvetica lumbersexual fixie. Direct trade gentrify chambray typewriter single-origin coffee wolf fam synth photo booth jean shorts hashtag man braid narwhal shaman. Woke etsy fixie hot chicken chia crucifix man bun mixtape.

###### Heading level six

Deep v seitan DSA, venmo yes plz big mood palo santo heirloom fingerstache craft beer cred normcore woke food truck chillwave. Godard tbh kickstarter, shabby chic ethical subway tile butcher coloring book pork belly. Skateboard authentic praxis wolf activated charcoal small batch hella pour-over big mood jean shorts brunch everyday carry pabst single-origin coffee. Marfa flannel thundercats yr, distillery JOMO hot chicken ethical.

## Heading level two

### Heading level three

I'm baby kale chips slow-carb cardigan gochujang, godard austin selvage master cleanse. Same four dollar toast af kinfolk, neutral milk hotel JOMO shabby chic sustainable hashtag ethical DSA raclette farm-to-table. Salvia chillwave vaporware Brooklyn. Gastropub DIY vinyl ascot stumptown williamsburg helvetica polaroid pitchfork tilde jawn kogi schlitz. Fam fanny pack glossier shoreditch occupy direct trade skateboard williamsburg godard scenester edison bulb salvia. Food truck viral meggings aesthetic, vaporware artisan organic twee before they sold out fingerstache mumblecore.

### Heading level three

#### Heading level four

I'm baby kale chips slow-carb cardigan gochujang, godard austin selvage master cleanse. Same four dollar toast af kinfolk, neutral milk hotel JOMO shabby chic sustainable hashtag ethical DSA raclette farm-to-table. Salvia chillwave vaporware Brooklyn. Gastropub DIY vinyl ascot stumptown williamsburg helvetica polaroid pitchfork tilde jawn kogi schlitz. Fam fanny pack glossier shoreditch occupy direct trade skateboard williamsburg godard scenester edison bulb salvia. Food truck viral meggings aesthetic, vaporware artisan organic twee before they sold out fingerstache mumblecore.

#### Heading level four

##### Heading level five

I'm baby kale chips slow-carb cardigan gochujang, godard austin selvage master cleanse. Same four dollar toast af kinfolk, neutral milk hotel JOMO shabby chic sustainable hashtag ethical DSA raclette farm-to-table. Salvia chillwave vaporware Brooklyn. Gastropub DIY vinyl ascot stumptown williamsburg helvetica polaroid pitchfork tilde jawn kogi schlitz. Fam fanny pack glossier shoreditch occupy direct trade skateboard williamsburg godard scenester edison bulb salvia. Food truck viral meggings aesthetic, vaporware artisan organic twee before they sold out fingerstache mumblecore.

##### Heading level five

###### Heading level six

I'm baby kale chips slow-carb cardigan gochujang, godard austin selvage master cleanse. Same four dollar toast af kinfolk, neutral milk hotel JOMO shabby chic sustainable hashtag ethical DSA raclette farm-to-table. Salvia chillwave vaporware Brooklyn. Gastropub DIY vinyl ascot stumptown williamsburg helvetica polaroid pitchfork tilde jawn kogi schlitz. Fam fanny pack glossier shoreditch occupy direct trade skateboard williamsburg godard scenester edison bulb salvia. Food truck viral meggings aesthetic, vaporware artisan organic twee before they sold out fingerstache mumblecore.
```

</details>

**Comment by caugner** [comment]
@pepelsbey Can you give an example for an MDN page to reproduce this?

**Comment by globau** [comment]
Putting this into the backlog to take off our radar; will re-evaluate once we get an example page.

**Comment by pepelsbey** [comment]
@caugner I updated the description with better structure and clearer description and also added a real example: [GLSL shaders](https://developer.mozilla.org/en-US/docs/Games/Techniques/3D_on_the_web/GLSL_Shaders#final_code). There are 4689 adjacent h2/h3 header occurrences in the content, mostly from example/demo pairs.

**Comment by caugner** [comment]
Thanks. We have this HTML structure, because Rari (and previously Yari) returns h2/h3 sections in one list (`doc.body`). See: https://developer.mozilla.org/en-US/docs/Games/Techniques/3D_on_the_web/GLSL_Shaders/index.json

@pepelsbey Can you clarify how this impacts users?

**Comment by pepelsbey** [comment]
> Can you clarify how this impacts users?

I did a quick test in VoiceOver, and the “Final code” heading describes an empty section, which might be confusing for non-sighted users. Specifically:

1. When you land on the section (or region, as VO calls it), it gets described by the heading (Final code, region)
2. But when you try to move forward to its content, it describes the section again and tells you “end of Final code, region”.

It makes it sound like the section is empty, which is not the case.
